homeassistant/components/climate/zha.py

```diff
@@ -0,0 +1,251 @@
+    """
+    try:
+        result, _ = yield from cluster.read_attributes(
+            attributes,
+            allow_cache=False,
+            raw=raw,
+        )
+        return result
+    except Exception:  # pylint: disable=broad-except
+        return {}
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the Zigbee Home Automation climate."""
+    discovery_info = zha.get_discovery_info(hass, discovery_info)
+    if discovery_info is None:
+        return
+
+    endpoint = discovery_info['endpoint']
+
+    try:
+        discovery_info['local_temp'] \
+            = yield from endpoint.thermostat['local_temp']
+    except (AttributeError, KeyError):
+        pass
+
+    # Clean data
+    for data in ('manufacturer', 'model'):
+        if isinstance(discovery_info.get(data), bytes):
+            discovery_info[data] = discovery_info[data]\
+                .split(b'\x00')[0].decode('utf-8')
+        elif isinstance(discovery_info.get(data), str):
+            discovery_info[data] = discovery_info[data].split('\x00')[0]
+
+    if discovery_info['manufacturer'] == 'Stelpro' and \
+            discovery_info['model'] == 'ST218':
+        async_add_devices([ClimateST218(**discovery_info)],
+                          update_before_add=True)
+    else:
+        _LOGGER.warning("ZigBee Climate %s %s not supported yet",
+                        discovery_info['manufacturer'],
+                        discovery_info['model'])
+
+
+class ClimateST218(zha.Entity, ClimateDevice):
+    """Climage Stelpro ST218."""
+
+    _domain = DOMAIN
+
+    def __init__(self, **kwargs):
+        """Initialize the ZHA Climate."""
+        super().__init__(**kwargs)
+
+        self._target_temperature = None
+        self._current_temperature = None
+        self._current_operation = None
+        self._heating_power = None
+        self.last_request = time.time()
+
+        # Add Stelpro Specific
+        import bellows.types as t
+        # Remove the following lines when this is merged
+        # https://github.com/rcloran/bellows/pull/71
+
+        class enum8(t.uint8_t):  # pylint: disable=C0103
+            """Missing bellows enum8 type."""
+
+            _enum = True
+
+        t.enum8 = enum8
+        from bellows.zigbee.zcl import foundation
+        foundation.DATA_TYPES[0x30] = ('Enumeration', t.enum8,
+                                       foundation.Discrete)
+        foundation.DATA_TYPE_IDX[t.enum8] = 0x30
+        self._endpoint.thermostat.attributes[0x001C] = \
+            ('system_mode', t.enum8)
+        # END
+
+        # pylint: disable=W0212
+        self._endpoint.thermostat._attridx['setpoint_mode_manuf_specific'] = \
+            0x401C
+        # pylint: enable=W0212
+        self._endpoint.thermostat.attributes[0x401C] = \
+            ('setpoint_mode_manuf_specific', t.enum8)
+
+    @asyncio.coroutine
+    def _wait_for_clean_communication(self):
```

Comment:
This probably fails if there are multiple requests "in the queue": Say I call a service that requires communication three times simultaneously. The first one goes through and sets `last_request` to the current time. The next two requests both sleep for one second and then perform their request within the same second (if the machine is fast enough). I think you need to loop here, calling `asyncio.sleep` until `time.time() - self.last_request >= CLEAN_TIME_ZONE` becomes true.

Reply:
Maybe you could add a lock that is acquired before entering the sleep and released after the sleep?